### ISSUE TITLE
S2E custom instructions are deprecated

### ISSUE THREAD

**Issue by riyadparvez** [stefanbucur/linux-s2e#1]
From https://github.com/stefanbucur/linux-s2e/blob/master/arch/x86/include/asm/s2e/s2e.h#L7:

```
/* (Deprecated) We define our own custom instruction, using a previously
 * unallocated opcode sequence. Therefore, this construct would cause a
 * processor exception on real hardware or non-S2E virtual environments.
 */
```

IIRC, in main S2E branch, there's nothing that said S2E custom instructions are deprecated. Can you please explain this?

**Comment by stefanbucur** [comment]
Yes, this code works against an experimental version of S2E that I'm currently working on. In this version, the S2E custom x86 opcode instructions have been replaced with overloaded x86 multi-byte NOP instructions, in order to allow the S2E-specific code to run on bare-metal hardware.

If you're interested, check out @ayekat's [S2E fork](https://github.com/ayekat/s2e) for a preview of this.

**Comment by riyadparvez** [comment]
Thanks for the reply.

Just out of curiosity, I understand the motivation behind overloading `nop` instructions. Is there any other benefit from symbolic execution point of view?

I'm working on a project on top of main branch of S2E. After publishing, I also intend to submit my contributions. I was wondering what's the probability of multi-bytes NOP getting integrated to upstream? As I'm also developing some guest APIs, so that I can target recent versions.

**Comment by stefanbucur** [comment]
> Just out of curiosity, I understand the motivation behind overloading nop instructions. Is there any other benefit from symbolic execution point of view?

This shouldn't affect in any way the symbolic execution itself. It's only the "protocol" between the guest and the S2E host that changes.

> I was wondering what's the probability of multi-bytes NOP getting integrated to upstream? 

I think it's very likely. The multi-byte NOPs simplify a lot of things in terms of dealing with guest binaries and I don't see any benefit to keeping the custom opcodes.
